server/controllers/user.js

```diff
@@ -0,0 +1,91 @@
+const User = require('../models/user')
+const bcrypt = require('bcryptjs')
+const jwt = require('../helpers/jwt')
+const { OAuth2Client } = require('google-auth-library');
+const client = new OAuth2Client(process.env.GOOGLE_CLIENT_KEY);
+
+class UserController {
+    static register(req, res, next) {
+        console.log(req.body);
+
+        User.create({
+            username: req.body.username,
+            email: req.body.email,
+            password: req.body.password
+        })
+            .then((newuser) => {
+                res.status(201).json(newuser)
+            })
+            .catch(next)
+    }
+
+    static login(req, res, next) {
+        console.log(req.body);
+
+        User.findOne({ email: req.body.email }, function (err, user) {
+            if (err) {
+                throw err
+            } else {
+                if (user) {
+                    if (bcrypt.compareSync(req.body.password, user.password)) {
+                        let obj = {
+                            id: user._id,
+                            email: user.email
+                        }
+                        res.json({ token: jwt.sign(obj) })
+
+                    } else {
+                        res.status(400).json({
+                            message: "wrong password"
+                        })
+                    }
+                } else {
+                    res.status(404).json({
+                        message: "username wrong"
+                    })
+                }
+            }
+        })
+    }
+
+    static signInWithGoogle(req, res, next) {
+        client
+            .verifyIdToken({
+                idToken: req.body.id_token,
+                audience: process.env.GOOGLE_CLIENT_KEY
+            })
+            .then(function (ticket) {
+                let accessToken
+                User.findOne({ email: email }, function (err, user) {
+                    if (err) {
+                        res.status(500).json({
+                            message: "Internal Server Error"
+                        })
+                    } else {
+                        if (user == null) {
+                            return User.create({
+                                username: `${ticket.getPayload().given_name}`,
+                                email: email,
+                                password: `todo${ticket.getPayload().given_name}`
+                            })
+                        } else {
+                            const { email } = ticket.getPayload();
+                            accessToken = jwt.sign({ email })
+                        }
+                    }
+                })
+
+                res.status(200).json({ accessToken });
+            })
+            .then((newuser) => {
+                console.log('create')
+                res.status(201).json(newuser)
```

Comment:
kan ini google sign in, jadi harusnya tetap ngebalikin token, bukan informasi user doang. karena pada hakikatnya 3rd party login ini fungsinya nge skip proses register, dan bisa masuk ke app kita tanpa user melakukan register/ memakai data account lain. jadi balikan proses ini mau user udah ada di database atau belum tetap access_token